docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,361 @@
+This will allow overriding the `jib.from.image` property that sets the base image.
+
+For Maven:
+
+```yaml
+build:
+  artifacts:
+  - image: base-image
+  - image: test-jib-maven
+    jib:
+      type: maven
+      args: 
+      - -Djib.from.image=docker://{{ .BASE_IMAGE }}
+    requires:
+      - image: base-image
+        alias: BASE_IMAGE
+```
+
+where `jib.from.image` is added appropriately the `pom.xml`
+
+Similarly, for Gradle:
+
+```yaml
+build:
+  artifacts:
+  - image: base-image
+  - image: test-jib-gradle
+    jib:
+      type: gradle
+      args: 
+      - -Djib.from.image=registry://{{ .BASE_IMAGE }}
+    requires:
+      - image: base-image
+        alias: BASE_IMAGE
+```
+
+where `jib.from.image` is added to the Gradle configuration.
+
+### Bazel builder
+
+The bazel builder doesn't support referencing images directly. Also, it natively supports setting up nested builds. We allow defining `required` artifacts even though they can't be referenced by the builder. We do this to give the user a way of ordering these builds; with the future work around [Skaffold Hooks](https://github.com/GoogleContainerTools/skaffold/issues/1441) there might be some usecases where pre and post build scripts might want a certain ordering of builds. 
+
+## Builder interfaces
+There are two builder abstractions -- in [build.go](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/build.go#L37) and [parallel.go](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/parallel.go#L34)
+
+```go
+type Builder interface {
+	Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+}
+```
+```go
+type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string) (string, error)
+```
+The first describes a builder for a list of artifacts. There are a few implementations like `cache`, `local` and `cluster` builders.
+The second describes a per artifact builder. Again there are a few implementations like `docker`, `buildpacks`, etc.
+
+We modify both of them to:
+
+```go
+type Builder interface {
+	Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, existing []Artifact) ([]Artifact, error)
+}
+```
+```go
+type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string, artifactResolver ArtifactResolver) (string, error)
+```
+where we define `ArtifactResolver` interface, as:
+```go
+// ArtifactResolver provides an interface to resolve built artifacts by image name.
+type ArtifactResolver interface {
+	GetImageTag(imageName string) string
+}
+```
+
+This necessitates all multi-artifact builder implementations to also accept a slice of already built artifacts' information. All single artifact builders require an `ArtifactResolver` that can provide the required artifacts.
```

Comment:
It would be helpful to explain why multi-artifact builders need a list of already-built artifacts (vs an ArtifactResolver)